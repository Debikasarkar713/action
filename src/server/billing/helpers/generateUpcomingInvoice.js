import fetchAllLines from 'server/billing/helpers/fetchAllLines';
import generateInvoice from 'server/billing/helpers/generateInvoice';
import stripe from 'server/billing/stripe';
import getRethink from 'server/database/rethinkDriver';
import resolvePromiseObj from 'universal/utils/resolvePromiseObj';
import getUpcomingInvoiceId from 'server/utils/getUpcomingInvoiceId';

const generateUpcomingInvoice = async (orgId) => {
  const r = getRethink();
  const invoiceId = getUpcomingInvoiceId(orgId);
  const {stripeId, stripeSubscriptionId} = await r.table('Organization').get(orgId)
    .pluck('stripeId', 'stripeSubscriptionId');
  const {stripeLineItems, upcomingInvoice} = await resolvePromiseObj({
    stripeLineItems: fetchAllLines('upcoming', stripeId),
    upcomingInvoice: stripe.invoices.retrieveUpcoming(stripeId, {subscription: stripeSubscriptionId})
  });
  return generateInvoice(upcomingInvoice, stripeLineItems, orgId, invoiceId);
};

export default generateUpcomingInvoice;
