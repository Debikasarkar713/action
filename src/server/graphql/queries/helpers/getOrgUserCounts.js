const getOrgUserCounts = (org) => {
  return org('orgUsers')
    .filter({inactive: true})
    .count()
    .default(0)
    .do((inactiveUserCount) => {
      return {
        orgUserCount: {
          activeUserCount: org('orgUsers').count().sub(inactiveUserCount),
          inactiveUserCount
        }
      };
    });
};

export default getOrgUserCounts;
