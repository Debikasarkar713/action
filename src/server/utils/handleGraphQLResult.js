import jwtDecode from 'jwt-decode';

const setSocketAuth = (socket, authTokenStr) => {
  socket.setAuthToken({
    ...jwtDecode(authTokenStr),
    exp: undefined
  });
};

const handleGraphQLResult = (result, socket) => {
  const {data, errors} = result;
  if (errors) return false;
  if (data.notificationsAdded) {
    const {notifications} = data.notificationsAdded;
    const updateTokenNotification = notifications.find((notification) => notification.authToken);
    if (updateTokenNotification) {
      setSocketAuth(socket, updateTokenNotification.authToken);
      return true;
    }
  } else if (data.acceptTeamInviteNotification) {
    setSocketAuth(socket, data.acceptTeamInviteNotification.authToken);
    return true;
  } else if (data.newAuthToken) {
    // TODO refactor the above 2 clauses so this is the only one
    setSocketAuth(socket, data.newAuthToken);
    return true;
  }
  return false;
};

export default handleGraphQLResult;
