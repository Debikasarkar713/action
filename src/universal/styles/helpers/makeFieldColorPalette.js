
import ui from 'universal/styles/ui';

export default function makeFieldColorPalette(colorPalette, hasPseudoClassStyles = true, customStyles) {
  const defaultCustomStyles = {
    base: {},
    placeholder: {},
    selection: {},
    hover: {},
    focus: {},
    active: {},
    ...customStyles
  };
  const {
    base,
    placeholder,
    selection,
    hover,
    focus,
    active
  } = defaultCustomStyles;
  const palettes = ui.fieldColorPalettes;
  const hoverFocusActive = {
    borderColor: palettes[colorPalette].focusBorderColor,
    boxShadow: ui.fieldFocusBoxShadow,
    outline: 'none'
  };
  const pseudoClassStyles = {
    ':hover': {
      ...hoverFocusActive,
      ...hover
    },
    ':focus': {
      ...hoverFocusActive,
      ...focus
    },
    ':active': {
      ...hoverFocusActive,
      ...active
    }
  };
  const addPsuedoClassStyles = hasPseudoClassStyles && pseudoClassStyles;
  return {
    backgroundColor: palettes[colorPalette].backgroundColor,
    borderColor: palettes[colorPalette].borderColor,
    borderRadius: ui.fieldBaseStyles.borderRadius,
    color: palettes[colorPalette].color,
    ...base,
    ...palettes[colorPalette].placeholder,
    ...placeholder,

    '::selection': {
      backgroundColor: palettes[colorPalette].selection,
      ...selection
    },

    ...addPsuedoClassStyles
  };
}
