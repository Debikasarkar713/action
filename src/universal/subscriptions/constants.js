// EVENTS
export const EDIT = 'EDIT';
export const LEAVE = 'LEAVE';
export const PRESENT = 'PRESENT';
export const SOUNDOFF = 'SOUNDOFF';

// CHANNELS
export const ARCHIVED_PROJECTS = 'archivedProjects';
export const AGENDA = 'agenda';
export const INVITATIONS = 'invitations';
export const ORG_APPROVALS = 'orgApprovals';
export const ORGANIZATIONS = 'organizations';
export const PRESENCE = 'presence';
export const PROJECTS = 'projects';
export const AGENDA_PROJECTS = 'agendaProjects';
export const TEAM = 'team';
export const TEAM_MEMBERS = 'teamMembers';
