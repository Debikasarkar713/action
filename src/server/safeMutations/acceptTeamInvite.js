import adjustUserCount from 'server/billing/helpers/adjustUserCount';
import getRethink from 'server/database/rethinkDriver';
import addUserToTMSUserOrg from 'server/safeMutations/addUserToTMSUserOrg';
import insertNewTeamMember from 'server/safeMutations/insertNewTeamMember';
import getTeamInviteNotifications from 'server/safeQueries/getTeamInviteNotifications';
import {auth0ManagementClient} from 'server/utils/auth0Helpers';
import getPubSub from 'server/utils/getPubSub';
import {ADD_USER} from 'server/utils/serverConstants';
import {
  ADD_TO_TEAM, JOIN_TEAM, NEW_AUTH_TOKEN, NOTIFICATIONS_ADDED,
  NOTIFICATIONS_CLEARED
} from 'universal/utils/constants';
import {getUserId} from 'server/utils/authorization';
import tmsSignToken from 'server/utils/tmsSignToken';

const acceptTeamInvite = async (teamId, authToken, email) => {
  const r = getRethink();
  const now = new Date();
  const userId = getUserId(authToken);
  const {team: {orgId, name: teamName}, user} = await r({
    team: r.table('Team').get(teamId).pluck('orgId', 'name'),
    user: r.table('User').get(userId)
  });
  const userOrgs = user.userOrgs || [];
  const userTeams = user.tms || [];
  const userInOrg = Boolean(userOrgs.find((org) => org.id === orgId));
  const tms = [...userTeams, teamId];
  const {expireInviteNotificationIds} = await r({
    // add the team to the user doc
    userUpdate: addUserToTMSUserOrg(userId, teamId, orgId),
    newTeamMember: insertNewTeamMember(userId, teamId),
    // find all possible emails linked to this person and mark them as accepted
    expireEmailInvitations: r.table('Invitation')
      .getAll(email, {index: 'email'})
      .update({
        acceptedAt: now,
        // flag the token as expired so they cannot reuse the token
        tokenExpiration: new Date(0),
        updatedAt: now
      }),
    expireInviteNotificationIds: getTeamInviteNotifications(orgId, teamId, [email])
      .delete({returnChanges: true})('changes')
      .map((change) => change('old_val')('id'))
      .default([])
  });

  if (!userInOrg) {
    await adjustUserCount(userId, orgId, ADD_USER);
  }
  auth0ManagementClient.users.updateAppMetadata({id: userId}, {tms});

  // Clear all invitiation notifications
  if (expireInviteNotificationIds.length > 0) {
    const notificationsCleared = {deletedIds: expireInviteNotificationIds};
    getPubSub().publish(`${NOTIFICATIONS_CLEARED}.${userId}`, {notificationsCleared});
  }

  // Tell the team who just joined
  const notificationsAdded = {
    notifications: [{
      type: JOIN_TEAM,
      teamName,
      preferredName: user.preferredName || user.email
    }]
  };
  getPubSub().publish(`${NOTIFICATIONS_ADDED}.${teamId}`, {notificationsAdded});

  // Send the new team member a welcome & a new token
  const newAuthToken = tmsSignToken(authToken, tms);
  const addedToTeam = {
    // no ID to rule out permanent notification. they already know, they are the ones who did it!
    authToken: newAuthToken,
    orgId,
    startAt: now,
    type: ADD_TO_TEAM,
    teamName,
    teamId
  };

  getPubSub().publish(`${NEW_AUTH_TOKEN}.${userId}`, {newAuthToken});
  getPubSub().publish(`${NOTIFICATIONS_ADDED}.${userId}`, {notificationsAdded: {notifications: [addedToTeam]}});
  return addedToTeam;
};

export default acceptTeamInvite;
