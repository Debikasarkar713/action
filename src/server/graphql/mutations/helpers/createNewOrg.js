import getRethink from 'server/database/rethinkDriver';
import {BILLING_LEADER, PERSONAL} from 'universal/utils/constants';

export default async function createNewOrg(orgId, orgName, leaderUserId) {
  const r = getRethink();
  const now = new Date();
  return r.table('Organization').insert({
    id: orgId,
    creditCard: {},
    createdAt: now,
    name: orgName,
    orgUsers: [{id: leaderUserId, role: BILLING_LEADER, inactive: false}],
    tier: PERSONAL,
    updatedAt: now
  }, {returnChanges: true})('changes')(0)('new_val');
}
