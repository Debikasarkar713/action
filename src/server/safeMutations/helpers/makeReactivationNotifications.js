import {REJOIN_TEAM} from 'universal/utils/constants';


const makeReactivationNotifications = (notifications, reactivatedUsers, teamMembers, inviter) => {
  const {teamName, userId} = inviter;
  const restOfTeamUserIds = teamMembers
    .filter((m) => m.isNotRemoved === true && m.id !== userId)
    .map((m) => m.userId);
  const notificationsToSend = {};
  reactivatedUsers.forEach((user, idx) => {
    const notification = notifications[idx];
    const {preferredName, id: reactivatedUserId} = user;

    // make a notification to the person being reactivated
    notificationsToSend[reactivatedUserId] = [notification];

    // make a notification for the other team members annoucing the reactivation
    const rejoinNotification = {
      teamName,
      preferredName,
      type: REJOIN_TEAM
    };
    restOfTeamUserIds.forEach((notificationUserId) => {
      if (inviter.userId === notificationUserId) return;
      notificationsToSend[notificationUserId] = notificationsToSend[notificationUserId] || [];
      notificationsToSend[notificationUserId].push(rejoinNotification);
    });
  });
  return notificationsToSend;
};

export default makeReactivationNotifications;
