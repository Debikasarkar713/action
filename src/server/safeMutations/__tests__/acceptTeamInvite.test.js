import DynamicSerializer from 'dynamic-serializer';
import MockDate from 'mockdate';
import MockDB from 'server/__tests__/setup/MockDB';
import {__now} from 'server/__tests__/setup/mockTimes';
import fetchAndSerialize from 'server/__tests__/utils/fetchAndSerialize';
import newInvitee from 'server/__tests__/utils/newInvitee';
import acceptTeamInvite from 'server/safeMutations/acceptTeamInvite';
import * as adjustUserCount from 'server/billing/helpers/adjustUserCount';
import {ADD_USER} from 'server/utils/serverConstants';
import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
import {NEW_AUTH_TOKEN, NOTIFICATIONS_ADDED, NOTIFICATIONS_CLEARED, TEAM_INVITE} from 'universal/utils/constants';
import {auth0ManagementClient} from 'server/utils/auth0Helpers';
import makeMockPubSub from 'server/__mocks__/makeMockPubSub';
import * as getPubSub from 'server/utils/getPubSub';
import getRethink from 'server/database/rethinkDriver';

MockDate.set(__now);
console.error = jest.fn();

describe('acceptTeamInvite', () => {
  beforeEach(() => {
    auth0ManagementClient.users.updateAppMetadata.mockReset();
  });

  test('adds an invitee who was not previously in the org', async () => {
    // SETUP
    adjustUserCount.default = jest.fn();
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const invitee = newInvitee('acceptTeamInvite1');
    await mockDB.init()
      .newUser({name: 'inviteeGuy', tms: [], userOrgs: [], email: invitee.email})
      .newNotification(undefined, {type: TEAM_INVITE, email: invitee.email});
    const teamId = mockDB.context.team.id;
    const inviteeUser = mockDB.context.user;
    const authToken = mockAuthToken(inviteeUser);
    const orgId = mockDB.context.organization.id;
    auth0ManagementClient.__initMock(mockDB.db);
    const mockPubSub = makeMockPubSub();
    getPubSub.default = () => mockPubSub;
    const tms = [teamId];
    // TEST
    const res = await acceptTeamInvite(teamId, authToken, inviteeUser.email);

    // VERIFY
    const db = await fetchAndSerialize({
      user: r.table('User').get(inviteeUser.id),
      teamMember: r.table('TeamMember').get(`${inviteeUser.id}::${teamId}`),
      notification: r.table('Notification').getAll(orgId, {index: 'orgId'}).orderBy('userIds').coerceTo('array'),
      invitation: r.table('Invitation').getAll(inviteeUser.email, {index: 'email'}).orderBy('tokenExpiration').coerceTo('array')
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();

    expect(mockPubSub.publish.mock.calls[0][0]).toEqual(`${NOTIFICATIONS_CLEARED}.${inviteeUser.id}`);
    expect(mockPubSub.publish.mock.calls[1][0]).toEqual(`${NOTIFICATIONS_ADDED}.${teamId}`);
    expect(mockPubSub.publish.mock.calls[2][0]).toEqual(`${NEW_AUTH_TOKEN}.${inviteeUser.id}`);
    expect(mockPubSub.publish.mock.calls[3][0]).toEqual(`${NOTIFICATIONS_ADDED}.${inviteeUser.id}`);
    expect(mockPubSub.publish.mock.calls[3][1].notificationsAdded.notifications[0]).toEqual(res);
    expect(adjustUserCount.default).toHaveBeenCalledWith(inviteeUser.id, orgId, ADD_USER);
    expect(auth0ManagementClient.users.updateAppMetadata).toHaveBeenCalledWith({id: inviteeUser.id}, {tms});
  });
  test('adds an invitee who was previously in the org', async () => {
    // SETUP
    adjustUserCount.default = jest.fn();
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    await mockDB.init()
      .newTeam({orgId: mockDB.context.organization.id})
      .newNotification(undefined, {type: TEAM_INVITE, email: mockDB.db.user[1].email});
    const teamId = mockDB.context.team.id;
    const inviteeUser = mockDB.db.user[1];
    const authToken = mockAuthToken(inviteeUser);
    const orgId = mockDB.context.organization.id;
    auth0ManagementClient.__initMock(mockDB.db);
    const mockPubSub = makeMockPubSub();
    getPubSub.default = () => mockPubSub;
    const tms = [mockDB.db.team[0].id, teamId];

    // TEST
    const res = await acceptTeamInvite(teamId, authToken, inviteeUser.email);

    // VERIFY
    const db = await fetchAndSerialize({
      user: r.table('User').get(inviteeUser.id),
      teamMember: r.table('TeamMember').get(`${inviteeUser.id}::${teamId}`),
      notification: r.table('Notification').getAll(orgId, {index: 'orgId'}).orderBy('userIds').coerceTo('array'),
      invitation: r.table('Invitation').getAll(inviteeUser.email, {index: 'email'}).orderBy('tokenExpiration').coerceTo('array')
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();

    expect(mockPubSub.publish.mock.calls[0][0]).toEqual(`${NOTIFICATIONS_CLEARED}.${inviteeUser.id}`);
    expect(mockPubSub.publish.mock.calls[1][0]).toEqual(`${NOTIFICATIONS_ADDED}.${teamId}`);
    expect(mockPubSub.publish.mock.calls[2][0]).toEqual(`${NEW_AUTH_TOKEN}.${inviteeUser.id}`);
    expect(mockPubSub.publish.mock.calls[3][0]).toEqual(`${NOTIFICATIONS_ADDED}.${inviteeUser.id}`);
    expect(mockPubSub.publish.mock.calls[3][1].notificationsAdded.notifications[0]).toEqual(res);
    expect(adjustUserCount.default).toHaveBeenCalledTimes(0);
    expect(auth0ManagementClient.users.updateAppMetadata).toHaveBeenCalledWith({id: inviteeUser.id}, {tms});
  });
});
