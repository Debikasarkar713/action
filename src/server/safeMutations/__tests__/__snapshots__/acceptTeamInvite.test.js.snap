// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`acceptTeamInvite adds an invitee who was not previously in the org 1`] = `
Object {
  "invitation": Array [],
  "notification": Array [],
  "teamMember": Object {
    "checkInOrder": 4,
    "email": "invitee@mockTeam+acceptTeamInvite1.co",
    "id": 3,
    "isCheckedIn": null,
    "isFacilitator": true,
    "isLead": false,
    "isNotRemoved": true,
    "picture": "https://placeimg.com/100/100/animals",
    "preferredName": "inviteeGuy",
    "teamId": 1,
    "userId": 0,
  },
  "user": Object {
    "cachedAt": 2017-01-01T07:00:00.000Z,
    "createdAt": 2017-01-01T07:00:00.000Z,
    "email": "invitee@mockTeam+acceptTeamInvite1.co",
    "emailVerified": false,
    "id": 0,
    "identities": Array [],
    "inactive": false,
    "lastLogin": 2017-01-01T07:00:00.000Z,
    "lastSeenAt": 2017-01-01T07:00:00.000Z,
    "name": "inviteeGuy",
    "picture": "https://placeimg.com/100/100/animals",
    "preferredName": "inviteeGuy",
    "tms": Array [
      1,
    ],
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userOrgs": Array [
      Object {
        "id": 2,
        "role": null,
      },
    ],
    "welcomeSentAt": 2017-01-01T07:00:00.000Z,
  },
}
`;

exports[`acceptTeamInvite adds an invitee who was previously in the org 1`] = `
Object {
  "invitation": Array [],
  "notification": Array [],
  "teamMember": Object {
    "checkInOrder": 6,
    "email": "vision@avengers.org",
    "id": 5,
    "isCheckedIn": null,
    "isFacilitator": true,
    "isLead": false,
    "isNotRemoved": true,
    "picture": "https://placeimg.com/100/100/animals",
    "preferredName": "Vision",
    "teamId": 2,
    "userId": 0,
  },
  "user": Object {
    "cachedAt": 2017-01-01T07:00:00.000Z,
    "createdAt": 2017-01-01T07:00:00.000Z,
    "email": "vision@avengers.org",
    "emailVerified": false,
    "id": 0,
    "identities": Array [],
    "inactive": false,
    "lastLogin": 2017-01-01T07:00:00.000Z,
    "lastSeenAt": 2017-01-01T07:00:00.000Z,
    "name": "Jonas",
    "picture": "https://placeimg.com/100/100/animals",
    "preferredName": "Vision",
    "tms": Array [
      1,
      2,
    ],
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userOrgs": Array [
      Object {
        "id": 3,
        "role": null,
      },
      Object {
        "id": 4,
        "role": null,
      },
    ],
    "welcomeSentAt": 2017-01-01T07:00:00.000Z,
  },
}
`;
